File: Juros.R
# Versão 0.2:    04/2024: trocada avaliação soZero por acumulador == 0
#             21/03/2026: corrigidos retornos com return() nos ifs; padronização de floats (.0);
#                         uso de ^ e || (em vez de ** e |)¹; pré-alocação de vetores¹;
#                         saída formatada com cat(sprintf())¹; ¹ = sugestões do ChatGPT

Juros <- list(quantidade = 0, composto = FALSE, periodo = 30, pagamentos = c(), pesos = c())

# calcula o peso total de todas as parcelas (números arbitrários)
getPesoTotal = function() {
  acumulador <- 0.0
  for (indice in 1:Juros$quantidade) {
    acumulador <- acumulador + Juros$pesos[indice]
  }
  acumulador
}

# calcula o acréscimo a partir de juros e parcelas
jurosParaAcrescimo = function(juros) {
  pesoTotal <- getPesoTotal()
  if (juros <= 0.0 || Juros$quantidade < 1 || Juros$periodo <= 0.0 || pesoTotal <= 0.0) { return(0.0) }

  acumulador <- 0.0

  for (indice in 1:Juros$quantidade) {
    if (Juros$composto) {
      acumulador <- acumulador + Juros$pesos[indice] / (1.0 + juros / 100.0) ^ (Juros$pagamentos[indice] / Juros$periodo)
    } else {
      acumulador <- acumulador + Juros$pesos[indice] / (1.0 + juros / 100.0 * Juros$pagamentos[indice] / Juros$periodo)
    }
  }
  
  if (acumulador <= 0.0) { return(0.0) }
  (pesoTotal / acumulador - 1.0) * 100.0
}

# calcula os juros a partir de acréscimo e parcelas
acrescimoParaJuros = function(acrescimo, precisao, maximoIteracoes, maximoJuros) {
  pesoTotal <- getPesoTotal()
  if (maximoIteracoes < 1 || Juros$quantidade < 1 || precisao < 1 || Juros$periodo <= 0.0 || acrescimo <= 0.0 || pesoTotal <= 0.0) { return(0.0) }
  minimoJuros <- 0.0
  medioJuros <- maximoJuros / 2.0
  minimaDiferenca <- 0.1 ^ precisao

  for (indice in 1:maximoIteracoes) {
    if ((maximoJuros - minimoJuros) < minimaDiferenca) { return(medioJuros) }
    if(jurosParaAcrescimo(medioJuros) <= acrescimo) {
      minimoJuros <- medioJuros 
    } else {
      maximoJuros <- medioJuros
    }
    medioJuros <- (minimoJuros + maximoJuros) / 2.0
  }
  
  medioJuros
}
  
# inicializa escalares
Juros$quantidade <- 3
Juros$composto <- TRUE
Juros$periodo <- 30.0

# inicializa listas
Juros$pagamentos <- numeric(Juros$quantidade)
Juros$pesos <- numeric(Juros$quantidade)

for(indice in 1:Juros$quantidade) {
  Juros$pagamentos[indice] = indice * Juros$periodo
  Juros$pesos[indice] = 1.0
}

# calcula e guarda os resultados das funções
pesoTotal <- getPesoTotal()
acrescimoCalculado <- jurosParaAcrescimo(3.0)
jurosCalculado <- acrescimoParaJuros(acrescimoCalculado, 15, 65, 50.0)

# imprime os resultados
cat(sprintf("Peso total = %.15f\n", pesoTotal))
cat(sprintf("Acréscimo = %.15f\n", acrescimoCalculado))
cat(sprintf("Juros = %.15f\n", jurosCalculado))

File: juros.rb
class Juros
  attr_accessor :quantidade, :composto, :periodo, :pagamentos, :pesos

  # construtor (perceba que o padrão de quantidade quebra a lógica dos cálculos)
  def initialize(quantidade = 0, composto = false, periodo = 30.0)
    @quantidade = quantidade
    @composto = composto
    @periodo = periodo
    @pagamentos = []
    @pesos = []
  end

  # retorna a soma de todos os pesos totais (obrigatório maior que zero para não dividir por zero)
  def getPesoTotal()
    acumulador = 0.0
    for indice in 0..@quantidade-1
      acumulador += @pesos[indice]
    end
    return acumulador
  end

  # calcula o acréscimo a partir dos juros e parcelas
  def jurosParaAcrescimo(juros)
    if juros <= 0 or @quantidade <= 0 or @periodo <= 0
      return 0.0
    end
    pesoTotal = getPesoTotal()
    if pesoTotal <= 0
      return 0.0
    end

    acumulador = 0.0
    soZero = true

    for indice in 0..@quantidade-1
      if @pagamentos[indice] > 0 and @pesos[indice] > 0
        soZero = false
      end
      if @composto
        acumulador += ( @pesos[indice] / (1.0 + juros / 100.0) ** (@pagamentos[indice] / @periodo) )
      else
        acumulador += ( @pesos[indice] / (1.0 + juros / 100.0 * @pagamentos[indice] / @periodo) )
      end
    end

    if soZero
      return 0.0
    end
    return (pesoTotal / acumulador - 1) * 100
  end

  # calcula os juros a partir do acréscimo e parcelas
  def acrescimoParaJuros(acrescimo, precisao = 15, maxIteracoes = 100, maxJuros = 50.0)
    if maxIteracoes < 1 or @quantidade <= 0 or precisao < 1 or @periodo <= 0 or acrescimo <= 0 or maxJuros <= 0
      return 0.0
    end
    minJuros = 0.0
    medJuros = maxJuros / 2
    minDiferenca = 0.1 ** precisao
    pesoTotal = getPesoTotal()
    if pesoTotal <= 0
      return 0.0
    end

    for indice in 1..maxIteracoes
      medJuros = (minJuros + maxJuros) / 2
      if (maxJuros - minJuros) < minDiferenca
        return medJuros
      end
      if jurosParaAcrescimo(medJuros) <= acrescimo
        minJuros = medJuros
      else
        maxJuros = medJuros
      end
    end

    return medJuros
  end
end

#testes
juros = Juros.new(3, true, 30.0)
juros.pagamentos[0] = 30.0
juros.pagamentos[1] = 60.0
juros.pagamentos[2] = 90.0
juros.pesos[0] = 1.0
juros.pesos[1] = 1.0
juros.pesos[2] = 1.0

puts(juros.getPesoTotal())
puts(juros.jurosParaAcrescimo(3.0))
puts(juros.acrescimoParaJuros( juros.jurosParaAcrescimo(3.0)))
